vlasovsolver/cpu_trans_pencils.hpp

```diff
@@ -0,0 +1,208 @@
+      N = 0;
+      sumOfLengths = 0;
+   }
+
+   void removeAllPencils() {
+      N = 0;
+      sumOfLengths = 0;
+      sumOfLengths = 0;
+      lengthOfPencils.clear();
+      idsStart.clear();
+      ids.clear();
+      sourceDZ.clear();
+      targetRatios.clear();
+      x.clear();
+      y.clear();
+      periodic.clear();
+      path.clear();
+   }
+
+   void addPencil(std::vector<CellID> idsIn, Real xIn, Real yIn, bool periodicIn, std::vector<uint> pathIn) {
+      N++;
+      // If necessary, add the zero cells to the beginning and end
+      if (idsIn.front() != 0) {
+         idsIn.insert(idsIn.begin(),0);
+         idsIn.insert(idsIn.begin(),0);
+      }
+      if (idsIn.back() != 0) {
+         idsIn.push_back(0);
+         idsIn.push_back(0);
+      }
+      sumOfLengths += idsIn.size();
+      lengthOfPencils.push_back(idsIn.size());
+      idsStart.push_back(ids.size());
+      ids.insert(ids.end(),idsIn.begin(),idsIn.end());
+      sourceDZ.resize(sumOfLengths);
+      targetRatios.resize(sumOfLengths);
+      x.push_back(xIn);
+      y.push_back(yIn);
+      periodic.push_back(periodicIn);
+      path.push_back(pathIn);
+   }
+
+   void removePencil(const uint pencilId) {
+      x.erase(x.begin() + pencilId);
+      y.erase(y.begin() + pencilId);
+      periodic.erase(periodic.begin() + pencilId);
+      path.erase(path.begin() + pencilId);
+
+      uint ibeg = idsStart[pencilId];
+      ids.erase(ids.begin() + ibeg, ids.begin() + ibeg + lengthOfPencils[pencilId] + 2*VLASOV_STENCIL_WIDTH);
+      targetRatios.erase(targetRatios.begin() + ibeg, targetRatios.begin() + ibeg + lengthOfPencils[pencilId] + 2*VLASOV_STENCIL_WIDTH);
+      sourceDZ.erase(sourceDZ.begin() + ibeg, sourceDZ.begin() + ibeg + lengthOfPencils[pencilId] + 2*VLASOV_STENCIL_WIDTH);
+      idsStart.erase(idsStart.begin() + pencilId);
+
+      N--;
+      sumOfLengths -= lengthOfPencils[pencilId];
+      lengthOfPencils.erase(lengthOfPencils.begin() + pencilId);
+   }
+
+   std::vector<CellID> getIds(const uint pencilId) const {
+      if (pencilId >= N) {
+         std::vector<CellID> idsEmpty;
+         return idsEmpty;
+      }
+      // Use vector range constructor. Only return actual pencil ids, not the stencils at the ends
+      std::vector<CellID>::const_iterator ibeg = ids.begin() + idsStart[pencilId] + VLASOV_STENCIL_WIDTH;
+      std::vector<CellID>::const_iterator iend = ibeg + lengthOfPencils[pencilId] - 2*VLASOV_STENCIL_WIDTH;
+      std::vector<CellID> idsOut(ibeg, iend);
+      return idsOut;
+   }
+
+   // Split one pencil into up to four pencils covering the same space.
+   // dx and dy are the dimensions of the original pencil.
+   void split(const uint myPencilId, const Real dx, const Real dy) {
+      auto myIds = this->getIds(myPencilId);
+
+      // Find paths that members of this pencil may have in other pencils (can happen)
+      // so that we don't add duplicates.
+      std::vector<int> existingSteps;
+
+#pragma omp parallel for
```

Comment:
No big difference with a schedule here I suppose?

Reply:
I would not schedule this. The cost is not suitable for static, but going to (1) would be overkill as it's still a very cheap loop, and not done often.